weather-app/src/app/search/search.component.ts

```diff
@@ -1,27 +1,81 @@
-import { Component, OnInit } from '@angular/core';
+import { Component, OnInit, ViewChild } from '@angular/core';
+import { FormControl } from '@angular/forms';
+import { MatAutocompleteTrigger } from '@angular/material/autocomplete';
+import { Router } from '@angular/router';
+import { debounceTime, distinctUntilChanged, map, Observable, of, startWith, switchMap } from 'rxjs';
+import { SearchAutocompleteService } from '../services/search-autocomplete.service';
 
 @Component({
   selector: 'app-search',
   templateUrl: './search.component.html',
   styleUrls: ['./search.component.scss']
 })
+
 export class SearchComponent implements OnInit {
+  @ViewChild(MatAutocompleteTrigger) autocomplete: MatAutocompleteTrigger = {} as MatAutocompleteTrigger;
+
   defaultCity = "Kyiv";
+  public myControl = new FormControl();
+  public filteredOptions: Observable<any[]> = new Observable();
+  public array = [];
+  public showNotFoundError = false;
+  public inputValue: string = '';
+
+  private isInputValid: boolean = false;
 
-  constructor() { }
+  constructor(private searchAutoCompleteService: SearchAutocompleteService, private router: Router) {
+    this.filteredOptions = this.myControl.valueChanges.pipe(
+      startWith(''),
+      debounceTime(400),
+      distinctUntilChanged(),
+      switchMap(val => {
+        return this.filter(val || '')
+      })
+    )
+  }
 
   ngOnInit(): void {
   }
 
+  filter(val: string): Observable<any[]> {
+    this.showNotFoundError = false;
+    this.isInputValid = false;
+    if (val.length < 3) {
+      return of([]);
+    }
+    return this.searchAutoCompleteService.getCities(val)
+      .pipe(
+        map((response: any) => {
+          const filteredOptions: any[] = response.filter(
+            (option: any) => option.name.toLowerCase().indexOf(val.toLowerCase()) === 0);
+
+          if (filteredOptions.length === 0) {
+            this.showNotFoundError = true;
+          }
+
+          if (filteredOptions.some(
+            (option: any) => option.name.toLowerCase() === this.inputValue.toLowerCase())) {
+              this.isInputValid = true;
```

Comment:
same here. when you want to assign some boolean values to variables it could be done by assigning the boolean result of the `if` condition:
`this.isInputValid = filteredOptions.some(
            (option: any) => option.name.toLowerCase() === this.inputValue.toLowerCase());`

Reply:
Fixed.